crates/agglayer-prover/src/executor.rs

```diff
@@ -0,0 +1,275 @@
+                    prover: Arc::new(network_prover),
+                    verification_key,
+                    timeout: config.network_prover.proving_timeout,
+                },
+            ))
+        } else {
+            None
+        };
+
+        let prover = CpuProver::new();
+        let (proving_key, verification_key) = prover.setup(ELF);
+
+        let local = Self::build_local_service(
+            config.cpu_prover.proving_timeout,
+            config.cpu_prover.max_concurrency_limit,
+            LocalExecutor {
+                prover: Arc::new(prover),
+                proving_key,
+                verification_key,
+                timeout: config.cpu_prover.proving_timeout,
+            },
+        );
+
+        Self { network, local }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Request {
+    pub(crate) initial_state: LocalNetworkState,
+    pub(crate) batch_header: MultiBatchHeader<Keccak256Hasher>,
+}
+
+impl From<Request> for SP1Stdin {
+    fn from(request: Request) -> Self {
+        let mut stdin = SP1Stdin::new();
+
+        stdin.write(&request.initial_state);
+        stdin.write(&request.batch_header);
+
+        stdin
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Response {
+    pub(crate) proof: SP1ProofWithPublicValues,
+}
+
+impl Service<Request> for Executor {
+    type Response = Response;
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        let network = if let Some(s) = self.network.as_mut() {
+            s.call(req.clone())
+        } else {
+            async { Err(Error::NetworkProverDisabled) }.boxed()
+        };
+
+        let mut local = self.local.clone();
+
+        let fut = async move {
+            match network.await {
+                Ok(res) => Ok(res),
+                Err(err) => {
+                    error!("Network prover failed: {:?}", err);
+
+                    local.ready().await?.call(req).await
+                }
+            }
+        };
+
+        Box::pin(fut)
+    }
+}
+
+#[derive(Clone)]
+struct LocalExecutor {
+    proving_key: SP1ProvingKey,
+    verification_key: SP1VerifyingKey,
+    prover: Arc<CpuProver>,
+    timeout: Duration,
+}
+
+impl Service<Request> for LocalExecutor {
+    type Response = Response;
+
+    type Error = Error;
+
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, _req: Request) -> Self::Future {
+        let prover = self.prover.clone();
+        let stdin = SP1Stdin::new();
+        let opts = ProofOpts {
+            timeout: Some(self.timeout),
+            ..Default::default()
+        };
```

Comment:
Is this timeout the same as the one specified for `TimeoutLayer`? If so, I think it is unlikely to ever kick in anyway because the `TimeoutLayer` will terminate it first because it is started a tiny bit earlier.